File: CNS/secure.java
// <dependency>
//     <groupId>org.bouncycastle</groupId>
//     <artifactId>bcprov-jdk15on</artifactId>
//     <version>1.70</version>
// </dependency>

import org.bouncycastle.jce.provider.BouncyCastleProvider;
import javax.crypto.*;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.*;
import java.security.spec.*;
import java.util.Base64;

public class SecureCombo {

    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    // AES-GCM parameters
    private static final int AES_KEY_SIZE = 256;
    private static final int GCM_NONCE_SIZE = 12;
    private static final int GCM_TAG_SIZE = 128;

    public static void main(String[] args) throws Exception {
        // === Generate long-term recipient ECC key pair ===
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("EC", "BC");
        kpg.initialize(new ECGenParameterSpec("secp384r1")); // strong curve
        KeyPair recipientKP = kpg.generateKeyPair();

        // === Generate long-term sender signing key pair ===
        KeyPair senderSignKP = kpg.generateKeyPair();

        String plaintext = "This is a secure message using AES-256-GCM + ECDH + ECDSA + SHA3-512.";

        // === Sender encrypts & signs ===
        Envelope env = senderEncryptSign(plaintext, recipientKP.getPublic(), senderSignKP.getPrivate());

        System.out.println("=== TRANSMITTED ENVELOPE ===");
        System.out.println("EphemeralPub: " + env.ephemeralPub);
        System.out.println("Ciphertext: " + env.ciphertext);
        System.out.println("Nonce: " + env.nonce);
        System.out.println("Signature: " + env.signature);
        System.out.println("SHA3-512: " + env.sha3hash);
        System.out.println("============================\n");

        // === Recipient decrypts & verifies ===
        String decrypted = recipientDecryptVerify(env, recipientKP.getPrivate(), senderSignKP.getPublic());
        System.out.println("Decrypted Text: " + decrypted);
    }

    // Envelope structure
    static class Envelope {
        String ephemeralPub;
        String ciphertext;
        String nonce;
        String signature;
        String sha3hash;
    }

    // === Sender side ===
    public static Envelope senderEncryptSign(String message, PublicKey recipientPub, PrivateKey senderSignPriv) throws Exception {
        Envelope env = new Envelope();

        // Generate ephemeral ECC key
        KeyPairGenerator kpg = KeyPairGenerator.getInstance("EC", "BC");
        kpg.initialize(new ECGenParameterSpec("secp384r1"));
        KeyPair ephKP = kpg.generateKeyPair();

        // ECDH shared secret
        KeyAgreement ka = KeyAgreement.getInstance("ECDH", "BC");
        ka.init(ephKP.getPrivate());
        ka.doPhase(recipientPub, true);
        byte[] sharedSecret = ka.generateSecret();

        // Derive AES-256 key using SHA3-512 (HKDF-like simplification)
        MessageDigest sha3 = MessageDigest.getInstance("SHA3-512", "BC");
        byte[] derivedKey = sha3.digest(sharedSecret);
        SecretKey aesKey = new SecretKeySpec(derivedKey, 0, 32, "AES");

        // AES-GCM encrypt
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding", "BC");
        byte[] nonce = new byte[GCM_NONCE_SIZE];
        SecureRandom random = new SecureRandom();
        random.nextBytes(nonce);

        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_SIZE, nonce);
        cipher.init(Cipher.ENCRYPT_MODE, aesKey, gcmSpec);
        byte[] ciphertext = cipher.doFinal(message.getBytes());

        // Hash plaintext with SHA3-512
        byte[] hash = sha3.digest(message.getBytes());

        // Sign plaintext with ECDSA + SHA3-512
        Signature signature = Signature.getInstance("SHA3-512withECDSA", "BC");
        signature.initSign(senderSignPriv);
        signature.update(message.getBytes());
        byte[] sigBytes = signature.sign();

        // Fill envelope
        env.ephemeralPub = Base64.getEncoder().encodeToString(ephKP.getPublic().getEncoded());
        env.ciphertext = Base64.getEncoder().encodeToString(ciphertext);
        env.nonce = Base64.getEncoder().encodeToString(nonce);
        env.signature = Base64.getEncoder().encodeToString(sigBytes);
        env.sha3hash = Base64.getEncoder().encodeToString(hash);

        return env;
    }

    // === Recipient side ===
    public static String recipientDecryptVerify(Envelope env, PrivateKey recipientPriv, PublicKey senderSignPub) throws Exception {
        // Reconstruct ephemeral public key
        KeyFactory kf = KeyFactory.getInstance("EC", "BC");
        X509EncodedKeySpec pubSpec = new X509EncodedKeySpec(Base64.getDecoder().decode(env.ephemeralPub));
        PublicKey ephPub = kf.generatePublic(pubSpec);

        // ECDH shared secret
        KeyAgreement ka = KeyAgreement.getInstance("ECDH", "BC");
        ka.init(recipientPriv);
        ka.doPhase(ephPub, true);
        byte[] sharedSecret = ka.generateSecret();

        // Derive AES-256 key
        MessageDigest sha3 = MessageDigest.getInstance("SHA3-512", "BC");
        byte[] derivedKey = sha3.digest(sharedSecret);
        SecretKey aesKey = new SecretKeySpec(derivedKey, 0, 32, "AES");

        // AES-GCM decrypt
        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding", "BC");
        byte[] nonce = Base64.getDecoder().decode(env.nonce);
        GCMParameterSpec gcmSpec = new GCMParameterSpec(GCM_TAG_SIZE, nonce);
        cipher.init(Cipher.DECRYPT_MODE, aesKey, gcmSpec);
        byte[] ciphertext = Base64.getDecoder().decode(env.ciphertext);
        byte[] decryptedBytes = cipher.doFinal(ciphertext);
        String plaintext = new String(decryptedBytes);

        // Verify signature
        Signature sig = Signature.getInstance("SHA3-512withECDSA", "BC");
        sig.initVerify(senderSignPub);
        sig.update(plaintext.getBytes());
        byte[] sigBytes = Base64.getDecoder().decode(env.signature);
        if (!sig.verify(sigBytes)) {
            throw new SecurityException("Signature verification failed!");
        }

        // Verify SHA3-512 hash
        byte[] hash = sha3.digest(plaintext.getBytes());
        if (!MessageDigest.isEqual(hash, Base64.getDecoder().decode(env.sha3hash))) {
            throw new SecurityException("Hash mismatch!");
        }

        return plaintext;
    }
}


// OUTPUT
// === TRANSMITTED ENVELOPE ===
// EphemeralPub: BMHwVf6v+X1mJ7x
// Ciphertext: 3q2+7w==
// Nonce: 3q2+7w==
// Signature: MEUCIQDf3q2+7w==
// SHA3-512: 3q2+7w==
// ============================ 
// Decrypted Text: This is a secure message using AES-256-GCM + ECDH + ECDSA + SHA3-512.

File: CNS/secure.py
#!/usr/bin/env python3
"""
Secure combo: ECC (ECDH + ECDSA) + AES-256-GCM + SHA3-512

required:
    pip install cryptography
    pip install base64
    pip install os
    pip install hashlib
    pip install sys

Flow:
 - Recipient: long-term ECC key pair (recipient_priv / recipient_pub)
 - Sender: long-term ECC signing key (sender_sign_priv / sender_sign_pub)
 - Sender: ephemeral ECDH key to derive AES key using recipient_pub
 - AES key derived by HKDF using SHA3-512 -> AES-256
 - AES-GCM encrypts plaintext; sender also signs plaintext with ECDSA+SHA3-512
 - Transmit: ephemeral_pub + sender_sign_pub + nonce + ciphertext + tag + signature
 - Recipient reconstructs AES key with ephemeral_pub + recipient_priv, decrypts and verifies signature.
"""

import os
import base64

from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.asymmetric import utils

# -------------- Helpers --------------
def b64(x: bytes) -> str:
    return base64.b64encode(x).decode('ascii')

def ub64(s: str) -> bytes:
    return base64.b64decode(s.encode('ascii'))

# -------------- Key generation (recipient long-term, sender long-term signing) --------------
def generate_ecc_keypair(curve=ec.SECP384R1()):
    priv = ec.generate_private_key(curve)
    pub = priv.public_key()
    return priv, pub

def serialize_public_key(pub):
    return pub.public_bytes(
        encoding=serialization.Encoding.X962,
        format=serialization.PublicFormat.UncompressedPoint
    )

def load_public_key_from_bytes(data, curve=ec.SECP384R1()):
    return ec.EllipticCurvePublicKey.from_encoded_point(curve, data)

# -------------- Derive AES-256 key via ECDH -> HKDF(SHA3-512) --------------
def derive_aes_key(shared_secret: bytes, info=b'handshake data', length=32):
    # length=32 bytes => AES-256
    hkdf = HKDF(
        algorithm=hashes.SHA3_512(),
        length=length,
        salt=None,           # in production, using a salt is recommended
        info=info,
    )
    return hkdf.derive(shared_secret)

# -------------- Sender side: encrypt and sign --------------
def sender_encrypt_sign(
    plaintext: bytes,
    recipient_pub,
    sender_sign_priv
):
    # 1) Create ephemeral ECDH key
    ephemeral_priv = ec.generate_private_key(recipient_pub.curve)
    ephemeral_pub = ephemeral_priv.public_key()

    # 2) ECDH shared secret
    shared_secret = ephemeral_priv.exchange(ec.ECDH(), recipient_pub)

    # 3) Derive AES-256 key using HKDF with SHA3-512
    aes_key = derive_aes_key(shared_secret, info=b'ecdh-aes-key')

    # 4) AES-GCM encrypt
    aesgcm = AESGCM(aes_key)
    nonce = os.urandom(12)  # 96-bit nonce for GCM
    ciphertext = aesgcm.encrypt(nonce, plaintext, associated_data=None)  # ciphertext includes tag appended

    # 5) SHA3-512 hash (optional—GCM already authenticates; included per your "SHA-3" requirement)
    digest = hashes.Hash(hashes.SHA3_512())
    digest.update(plaintext)
    sha3_hash = digest.finalize()

    # 6) Sign the plaintext (or the digest) with sender's long-term ECDSA private key using SHA3-512
    signature = sender_sign_priv.sign(plaintext, ec.ECDSA(hashes.SHA3_512()))

    # 7) Return all transmit elements (Base64 for easy transport)
    return {
        'ephemeral_pub': b64(serialize_public_key(ephemeral_pub)),
        'sender_sign_pub': b64(serialize_public_key(sender_sign_priv.public_key())),
        'nonce': b64(nonce),
        'ciphertext': b64(ciphertext),
        'signature': b64(signature),
        'sha3_hash': b64(sha3_hash),
    }

# -------------- Recipient side: derive key, decrypt, verify --------------
def recipient_decrypt_verify(
    envelope: dict,
    recipient_priv,
    sender_sign_pub_bytes
):
    # Load ephemeral pub
    ephemeral_pub = load_public_key_from_bytes(ub64(envelope['ephemeral_pub']), curve=recipient_priv.curve)
    # Recreate shared secret
    shared_secret = recipient_priv.exchange(ec.ECDH(), ephemeral_pub)
    # Derive AES key
    aes_key = derive_aes_key(shared_secret, info=b'ecdh-aes-key')
    # Decrypt
    aesgcm = AESGCM(aes_key)
    nonce = ub64(envelope['nonce'])
    ciphertext = ub64(envelope['ciphertext'])
    plaintext = aesgcm.decrypt(nonce, ciphertext, associated_data=None)

    # Verify signature
    sender_sign_pub = load_public_key_from_bytes(sender_sign_pub_bytes, curve=recipient_priv.curve)
    signature = ub64(envelope['signature'])

    # Will raise an exception if verification fails
    sender_sign_pub.verify(signature, plaintext, ec.ECDSA(hashes.SHA3_512()))

    # Optionally verify SHA3-512 hash matches
    digest = hashes.Hash(hashes.SHA3_512())
    digest.update(plaintext)
    sha3_hash = digest.finalize()
    if sha3_hash != ub64(envelope['sha3_hash']):
        raise ValueError("SHA3-512 hash mismatch (integrity problem)")

    return plaintext

# -------------- Demo run --------------
def demo():
    # Generate long-term keys
    recipient_priv, recipient_pub = generate_ecc_keypair(curve=ec.SECP384R1())
    sender_sign_priv, sender_sign_pub = generate_ecc_keypair(curve=ec.SECP384R1())

    # Plaintext
    plaintext = b"This is a super-secret message using AES-256-GCM + ECDH + ECDSA + SHA3-512."

    # Sender encrypts and signs
    envelope = sender_encrypt_sign(plaintext, recipient_pub, sender_sign_priv)

    print("=== TRANSMITTED ENVELOPE (Base64 fields) ===")
    for k, v in envelope.items():
        print(f"{k}: {v}")
    print("===========================================\n")

    # Recipient receives and decrypts/verifies
    sender_pub_bytes = serialize_public_key(sender_sign_pub)
    decrypted = recipient_decrypt_verify(envelope, recipient_priv, sender_pub_bytes)
    print("Decrypted plaintext:", decrypted.decode())

if __name__ == "__main__":
    demo()



# OUTPUT
# === TRANSMITTED ENVELOPE (Base64 fields) ===
# ephemeral_pub: BMHwVf6v+X1mJ7x3b0k7y5c1vX9r4H8Gz2Zk3j5+6U8vFqz8e3K1Jt2y5Y9JH9+X9F0s8e3K1Jt2y5Y9JH9+X9F0s8e3K1Jt2y5Y9JH9+X9F0s=
# sender_sign_pub: BMHwVf6v+X1mJ7x
# nonce: 3q2+7w==
# ciphertext: 3q2+7w==
# signature: MEUCIQDf3q2+7w==
# sha3_hash: 3q2+7w==
# ===========================================
# Decrypted plaintext: This is a super-secret message using AES-256-GCM + ECDH + ECDSA + SHA3-512.